apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts

```diff
@@ -99,68 +87,48 @@ export class DeclarationReferenceGenerator {
+    const sourceFile: ts.SourceFile | undefined = declaration?.getSourceFile();
     const parent: ts.Symbol | undefined = TypeScriptInternals.getSymbolParent(symbol);
-    // First, try to determine navigation to symbol via its parent.
-    if (parent) {
-      if (
-        parent.exports &&
-        DeclarationReferenceGenerator._isSameSymbol(parent.exports.get(symbol.escapedName), symbol)
-      ) {
-        return Navigation.Exports;
-      }
-      if (
-        parent.members &&
-        DeclarationReferenceGenerator._isSameSymbol(parent.members.get(symbol.escapedName), symbol)
-      ) {
-        return Navigation.Members;
-      }
-      if (
-        parent.globalExports &&
-        DeclarationReferenceGenerator._isSameSymbol(parent.globalExports.get(symbol.escapedName), symbol)
-      ) {
-        return 'global';
-      }
+
+    // If it's a global, then use an Exports navigation.
+    if (sourceFile && !ts.isExternalModule(sourceFile)) {
+      return Navigation.Exports;
     }
 
-    // Next, try determining navigation to symbol by its node
-    if (symbol.valueDeclaration) {
-      const declaration: ts.Declaration = ts.isBindingElement(symbol.valueDeclaration)
-        ? ts.walkUpBindingElementsAndPatterns(symbol.valueDeclaration)
-        : symbol.valueDeclaration;
-      if (ts.isClassElement(declaration) && ts.isClassLike(declaration.parent)) {
-        // class members are an "export" if they have the static modifier.
-        return ts.getCombinedModifierFlags(declaration) & ts.ModifierFlags.Static
-          ? Navigation.Exports
-          : Navigation.Members;
-      }
-      if (ts.isTypeElement(declaration) || ts.isObjectLiteralElement(declaration)) {
-        // type and object literal element members are just members
-        return Navigation.Members;
-      }
-      if (ts.isEnumMember(declaration)) {
-        // enum members are exports
-        return Navigation.Exports;
-      }
-      if (
-        ts.isExportSpecifier(declaration) ||
-        ts.isExportAssignment(declaration) ||
-        ts.isExportSpecifier(declaration) ||
-        ts.isExportDeclaration(declaration) ||
-        ts.isNamedExports(declaration)
-      ) {
-        return Navigation.Exports;
+    // If it's from an external library, then use an Exports navigation.
+    if (sourceFile && this._collector.program.isSourceFileFromExternalLibrary(sourceFile)) {
+      return Navigation.Exports;
+    }
+
+    // Otherwise, this symbol is from the current package.
+    if (parent) {
+      // If we've found an exported CollectorEntity, then it's exported from the package entry point, so
+      // use an Exports navigation.
+      const namedDeclaration: ts.DeclarationName | undefined = (
+        declaration as ts.NamedDeclaration | undefined
+      )?.name;
+      if (namedDeclaration && ts.isIdentifier(namedDeclaration)) {
+        const collectorEntity: CollectorEntity | undefined =
+          this._collector.tryGetEntityForNode(namedDeclaration);
```

Comment:
The method `tryGetEntityForNode` throws if it's called on an identifier that hasn't been analyzed. Presumably this shouldn't happen here, because the only identifiers that are not analyzed are globals and external package symbols, which are both handled earlier.

Reply:
The `tryGetEntityForNode()` exception was a sanity check, to catch cases where the analyzer accidentally crawls outside the scope of declarations that were scanned by the earlier step.

Reply:
Are we worried at all about this method throwing now? Previously, the worst-case scenario of this method is it spits out an incorrect navigation step, which messes up any canonical references that use that step, but API Extractor still completes. Now, the worst-case scenario of this method is it throws. I don't have 100% confidence that this will never throw - does it make sense to wrap it in a try/catch or should we add a new method to the collector that lets us get an entity for a node without throwing?

Reply:
No, for this kind of complex logic, I think we want to know when it is broken.  If we don't report an error, nobody will report a GitHub issue.